torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1040,7 +1064,7 @@ def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
             for atom in expr.atoms(FloorDiv):
                 base, divisor = atom.args
                 if self.replace(base % divisor) in self.divisible:
-                    div_replacements[atom] = base / divisor
+                    div_replacements[atom] = sympy.floor(base / divisor)
```

Comment:
This is important to enable us to optimize:

```
Eq(FloorDiv(integer_expr1, sympy.floor(base/divisor), integer_expr2)
```

If we had this instead, like before:

```
Eq(FloorDiv(integer_expr1, base/divisor, integer_expr2)
```

It would return `False` immediately since the result of `FloorDiv` is an integer only when both args are integers (based on the defined assumptions).

Test that can be used to repro:
```
pytest test/test_proxy_tensor.py -k test_make_fx_symbolic_exhaustive_nn_functional_glu_cpu_float32 -v
```

Reply:
The problem looks legitimate but I think a better solution is to only learn about divisibility when yhe arguments are integer (and not try to do anything clever with float div.) would that work?

Reply:
Updated. The reason for not using assumptions there and using isinstance(..., Integer):

```
# works for Integer type
[ins] In [24]:  (sympy.Integer(42) / 2).is_integer
Out[24]: True

# this is FloorDiv, might call floor for complex exprs
[ins] In [25]:  (sympy.Integer(42) // 2).is_integer
Out[25]: True

# this is FloorDiv, might call floor for complex exprs
[ins] In [26]:  (sympy.Symbol("s", integer=True) // 2).is_integer
Out[26]: True

# None, doesn't work for integer Symbols
[ins] In [27]:  (sympy.Symbol("s", integer=True) / 2).is_integer
```

Reply:
Hmm, now that I think of it... Since we allow None in FloorDiv assumptions it should just fail to simplify with `/` for Symbols right away and eval later. Let me do more tests.

Reply:
Ah no, using assumptions in simplify breaks this:

```
python benchmarks/dynamo/timm_models.py --ci --training --accuracy \ 
  --dynamic-shapes --device cuda --backend aot_eager \
  --total-partitions 2 --partition-id 0 \
  --output log.txt -k eca_halonext26ts

```

```
RuntimeError: Failed running call_function <built-in function matmul>(*(FakeTensor(FakeTensor(..., device='meta',
           size=((s2**2*s0)//(8*(s2//8)**2), (s2//8)**2, (s2//(s2//8))*((s2*((8*s0)//((s2**2*s0)//(8*(s2//8)**2))))//(s2//8)), 16),
           grad_fn=<TransposeBackward0>), cuda:0), FakeTensor(FakeTensor(..., device='meta', size=(8*s0, (s2//8)**2, 16, 144),
           grad_fn=<TransposeBackward0>), cuda:0)), **{}):
expand: attempting to expand a dimension of length 8*s0!
```

This is the same issue that we saw before: the assumptions system is being too eager, which contradicts floordiv assumptions. So let's not do that.

Reply:
... but checking against Integer doesn't make sense here because we already have an Integer rule in FloorDiv, so it should eval automatically and this branch would never be executed :(

Reply:
@ezyang I don't think there's a way to have replacements unless we use `sympy.floor(base / divisor)` in simplify. The concrete example is https://github.com/pytorch/pytorch/pull/90906#issuecomment-1403061281 and `s2 / 8` is what we need to floor to make it an integer for FloorDiv + Eq to work.

Can I change this back to `sympy.floor(base / divisor)`? Or do you see another solution?

Reply:
woof. OK, I understand why you want the floor, but I'm just very afraid the floor will make the expressions worse. I think we'll need to consult with @Chillee on this.

Reply:
We've discussed this offline. Horace suggested to try defining a custom class that would do truediv, but have assumptions necessary to prove that `s2 / 8` is an integer no matter what. And then we can use it here instead of floor. So let me try that.

Reply:
@ezyang @Chillee I've attempted the "clean division" class idea. Unfortunately, I think it doesn't work:

1. `_eval_*` assumption handlers seem to be executed after `eval` and _only if_ the latter returns `None`
2. If you return `None` from `eval`, your new class becomes a node in all exprs, so hard to optimize (same issue as `floor`)
3. Additional maintenance burden from having this custom class (while `floor` captures the semantics of the operation perfectly) + there are some issues with handling it in `_maybe_evaluate_static`, I'm not sure why yet.

Now, let's go back to why we want to avoid `floor` here. IIUC, it's only because it's hard to optimize exprs containing it (+ that inductor bug, but that's separate). Is there anything else?

If it's only optimization, I think we can just keep `floor` and define custom rules to optimize it instead. E.g., Horace suggested that we might want to optimize cases like `floor(x / y) * y`. This is doable by overloading `Mul` and having `floor` elimination logic there. NB: `x` must be divisible by `y` for this to work, counterexample: `sympy.floor(sympy.Integer(7) / 3) * 3 != 7`.

TL;DR: I propose we merge this with `floor` because it passes the tests and the code is clean. Then we can discuss separately how to optimize cases we could be optimizing.

Please LMK if I can merge this or if I need to spend more time attempting to make this work without `floor`.

Reply:
OK, while Horace is coming up with a new suggestion, here's what I propose: let's keep the floor, conditional on it not regressing any inductor inference models in CI (we recently added these to CI, so rebase and turn on ciflow/inductor). Assuming no regressions, let's just ship it.

Reply:
> If you return None from eval, your new class becomes a node in all exprs, so hard to optimize

We can just define new simplification rules right? I don't think it's that difficult.

> If it's only optimization

IMO, my main objection is that philosophically, `floor(s0/4)` just seems ... wrong, if we know that s0 is divided cleanly by 4. It makes more sense to me to just have a `CleanDiv` object.

> Assuming no regressions, let's just ship it.

If there are no regressions, then I'm fine with it. But I somewhat suspect there will be regressions.